File: Graphs/graphs_bfs.go
// Breadth First Search
/*
	In this implementation, we define a Graph struct that represents a graph with a given number of vertices,
	edges, and a visited array to keep track of visited vertices. We also define two methods - addEdge to
	add an edge to the graph and BFS to perform the BFS algorithm.

	In the BFS method, we start by creating a queue to store the vertices to visit. We mark the start node
	as visited and enqueue it. Then, while the queue is not empty, we dequeue a vertex from the queue,
	print it out, and get all its adjacent vertices. For each adjacent vertex, if it hasn't been visited,
	we mark it as visited and enqueue it. This continues until all reachable vertices have been visited.

	In the main function, we create a new Graph with 5 vertices and add edges to it. We then perform
	BFS starting from vertex 2, and print out the visited vertices. The output of this program will be:
	BFS starting from vertex 2:
	2 0 3 1

	The time complexity of BFS (Breadth-First Search) algorithm is O(V + E), where V is the number of
	vertices and E is the number of edges in the graph. This is because BFS traverses all the vertices
	and edges of the graph exactly once, and the time taken to visit each vertex and edge is constant.
	Therefore, the time complexity of BFS is proportional to the size of the graph.

	The space complexity of BFS is O(|V|), where |V| is the number of vertices in the graph.
	This is because in the worst case scenario, we would need to store all vertices in the queue
	before we finish traversing the graph.
*/
package main

import "fmt"

// Define a Graph struct to represent a graph
type Graph struct {
	vertices int        // Number of vertices in the graph
	edges    [][]int    // Adjacency list to store edges
	visited  []bool     // Track if a vertex is visited or not
}

// Function to add an edge to the graph
func (g *Graph) addEdge(u, v int) {
	// Add edge from u to v
	g.edges[u] = append(g.edges[u], v)
	// Add edge from v to u
	g.edges[v] = append(g.edges[v], u)
}

// Function to perform Breadth First Search
func BFS(g *Graph, start int) {
	// Create a queue for BFS
	queue := []int{}
	// Mark the start node as visited and enqueue it
	g.visited[start] = true
	queue = append(queue, start)

	for len(queue) != 0 {
		// Dequeue a vertex from the queue
		currVertex := queue[0]
		queue = queue[1:]
		fmt.Printf("%d ", currVertex)

		// Get all adjacent vertices of the dequeued vertex currVertex
		// If an adjacent vertex has not been visited, mark it as visited and enqueue it
		for _, adjVertex := range g.edges[currVertex] {
			if !g.visited[adjVertex] {
				g.visited[adjVertex] = true
				queue = append(queue, adjVertex)
			}
		}
	}
}

func main() {
	// Create a new graph with 5 vertices
	g := Graph{
		vertices: 5,
		edges:    make([][]int, 5),
		 // Initially all vertices are unvisited
		visited:  make([]bool, 5),
	}

	// Add edges to the graph
	g.addEdge(0, 1)
	g.addEdge(0, 2)
	g.addEdge(1, 2)
	g.addEdge(2, 0)
	g.addEdge(2, 3)
	g.addEdge(3, 3)

	// Perform BFS starting from vertex 2
	fmt.Println("BFS starting from vertex 2:")
	BFS(&g, 2)
}

File: Graphs/Graphs_bfs.js
// Define a function that takes a graph and a starting node as input
function bfs(graph, startNode) {
    // Initialize an empty object to keep track of visited nodes
    const visited = {}
    // Initialize a queue with the starting node
    const queue = [startNode]
  
    // Mark the starting node as visited
    visited[startNode] = true
  
    // While there are nodes in the queue
    while (queue.length > 0) {
      // Get the next node from the front of the queue
      const currentNode = queue.shift()
      // Log the current node to the console (or do something else with it)
      console.log(currentNode)
  
      // Get the adjacent nodes of the current node from the graph
      const adjacentNodes = graph[currentNode]
  
      // For each adjacent node
      for (let i = 0; i < adjacentNodes.length; i++) {
        // Get the adjacent node
        const adjacentNode = adjacentNodes[i].node
  
        // If the adjacent node has not been visited
        if (!visited[adjacentNode]) {
          // Mark the adjacent node as visited
          visited[adjacentNode] = true
          // Add the adjacent node to the back of the queue
          queue.push(adjacentNode)
        }
      }
    }
  }

// SAMPLE USE CASE
// The graph will be represented in form of an adjacency list.
/* 
    The graph will be in form of an object where each key represents the nodes and each value will be an array of the neighbors of the node.

    The array in values will be an object array, where each object has the node, which is the neighbor, and weight, which is the distance to that neighbor from the current node.
*/
const graph = {
    0: [{ node: 1, weight: 10 }],
    1: [
        { node: 0, weight: 10 },
        { node: 2, weight: 10 },
        { node: 3, weight: 5 },
    ],
    2: [{ node: 1, weight: 10 }],
    3: [
        { node: 1, weight: 5 },
        { node: 4, weight: 10 },
        { node: 5, weight: 10 },
    ],
    4: [
        { node: 3, weight: 10 },
        { node: 5, weight: 10 },
    ],
    5: [
        { node: 3, weight: 10 },
        { node: 4, weight: 10 },
    ],
}

bfs(graph, 2)